src/common/k8s-api/api-manager/auto-registration.injectable.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getInjectable } from "@ogre-tools/injectable";
+import { KubeApi } from "../kube-api";
+import { KubeObject } from "../kube-object";
+import autoRegistrationEmitterInjectable from "./auto-registration-emitter.injectable";
+import apiManagerInjectable from "./manager.injectable";
+import { CustomResourceStore } from "./resource.store";
+
+const autoRegistrationInjectable = getInjectable({
+  id: "api-manager-auto-registration",
+  instantiate: (di) => {
+    const apiManager = di.inject(apiManagerInjectable);
+    const autoRegistrationEmitter = di.inject(autoRegistrationEmitterInjectable);
+
+    autoRegistrationEmitter
+      .on("customResourceDefinition", (crd) => {
+        const objectConstructor = class extends KubeObject {
+          static readonly kind = crd.getResourceKind();
+          static readonly namespaced = crd.isNamespaced();
+          static readonly apiBase = crd.getResourceApiBase();
+        };
+
+        const api = (() => {
+          const rawApi = apiManager.getApi(objectConstructor.apiBase);
+
+          if (rawApi) {
+            return rawApi;
+          }
+
+          const api = new KubeApi({ objectConstructor });
+
+          apiManager.registerApi(api);
```

Comment:
I have a feeling that code similar to this could benefit significantly from the Hollywood Principle, i.e. "Don't call us, we'll call you". That would mean that instead of us doing mutations, such as registrations, adds, and whatnot, e.g. `apiManager` could simply "call" for all the API's available when required, without the precondition of the API's calling in to register before that.

Note: Hollywood is not one of the SOLID principles, and possibly not as big a deal.